### PR TITLE
Bugfix/remove sneaky optimization and replace with virtual functions

### LINKED ISSUE #472: Loop on waveform view not working
### Please describe the problem:

In waveform view loop shortcut ( green / red bar + any pad ) doesn’t not working .

### What is the expected behavior?

Make loop section

### Is there a relevant Pull request?

I don’t know 

### What hardware did you reproduce it with?

OLED

### What firmware did you reproduce it with?

nightly

### What is the firmware name:

Deluge-v1_0_0-nightly+2023_08_29-dbaa842.bin

### If possible provide the steps to reproduce the issue and upload additional media:

1.open waveform view

2.press green bar+ any pad

result: nothing is happens

### LINKED ISSUE #492: Sample Preview shows incorrect waveform, file name, and audio playback isn't full length
### Please describe the problem:

With PLAY turned OFF (no green light)
With PREView set to ON, when browsing samples (audio files, .wav):
- On each new sample preview when turning select encoder to the right, the new sample audio might play for only half a second, the waveform displays either no waveform or the last correctly rendered waveform, not the correct waveform for the sample, and the file name on 7SEG displays the last correctly previewed sample and doesn't change when scrolling
- By switching back and forth between two samples in a folder, eventually the sample will playback for the full length correctly, eventually the correct waveform for the audio is rendered
- Eventually all samples in a folder can playback correctly with enough preview attempts
- Samples that might be previewing correctly will go back to being incorrect after switching to other folders and previewing other samples (unsure, hard to verify)

With PLAY turned ON (green light!)
- This behavior seems much less frequent if there at all
- A folder of incorrectly previewing samples seem to start previewing correctly
- Related to Issue #350, quickly scrolling samples while previewing can cull voices a lot quicker than they would usually be culled. 

It seems inconsistent but reproducible, tested on Cordae song and another song, two different nightly firmwares, multiple different folders of samples of varying lengths. 

### What is the expected behavior?

If PREView is set to ON in the System Settings, when turning the select knob to select an audio file, or when entering into a folder of audio files (.wav):
- it plays the audio file from start to finish
- the waveform for the current sample that is playing is displayed on the main grid pads
- its file name is displayed on the display

### Is there a relevant Pull request?

unknown

### What hardware did you reproduce it with?

7SEG

### What firmware did you reproduce it with?

nightly

### What is the firmware name:

deluge-v1_0_0-nightly+2023_08_31-b4aeb97

### If possible provide the steps to reproduce the issue and upload additional media:

To reproduce, browse into a folder of samples, turn select encoder to the right and for each click of the encoder, listen for the preview playback to be cut out after a short amount of time, notice if the waveform displayed is the same as the previous sample previewed, and notice if the file name displayed is the same as the previous sample previewed.

If an incorrect previewed sample is found, turn select left and right one time each to confirm if that sample isn't loading correctly. Attempt previewing the sample again and again until eventually the sample starts to preview correctly. It takes usually 2-5 attempts to get a sample to preview correctly. 

Recorded a video here, hopefully the link works, file was too big to upload: https://photos.app.goo.gl/Ehmias3AUy92z3U57

### LINKED ISSUE #610: Scrolling through synths - sample based synths don't always load.
### Please describe the problem:

When choosing a synth for a track, sample based synths don't always load.  Sometimes I have to go past it and then scroll back to get it to load.  This is on OLED with community build beta.  It's pretty consistent in not loading most of the time.

### What is the expected behavior?

Sample based synths should load just like regular synths.

### Is there a relevant Pull request?

_No response_

### What hardware did you reproduce it with?

OLED

### What firmware did you reproduce it with?

community

### What is the firmware name:

deluge-v1_0_0-beta+2023_10_02-d9fbde6.bin

### If possible provide the steps to reproduce the issue and upload additional media:

Use scroll knob to browse various saved synths on a track.  Name will show, but no response to trigger input via midi keyboard or audition buttons.  Scroll a couple of synths past and then back again usually fixes it.  Mostly going past and back.  Going before and back usually doesn't work.

### PR DESCRIPTION
**Pull request by m-m-adams**
The sneaky optimization seems to have resulted in issues around multi samples (which I think is actually all samples internally)

With this fix I can no longer reproduce the following bugs:

Fix #610 
Fix #492
~~Fix #472~~ still happens sometimes

### CODE REVIEW

File: src/deluge/storage/multi_range/multisample_range.h [hunk 1]
```diff
@@ -29,6 +29,7 @@ class Cluster;
 class MultisampleRange final : public MultiRange {
 public:
 	MultisampleRange();
+	AudioFileHolder* getAudioFileHolder();
```
Comment (bobtwinkles) [inline src/deluge/storage/multi_range/multisample_range.h:32]: Should have an `override` at the end here

File: src/deluge/storage/multi_range/multi_wave_table_range.h [hunk 2]
```diff
@@ -23,6 +23,7 @@
 class MultiWaveTableRange final : public MultiRange {
 public:
 	MultiWaveTableRange();
+	AudioFileHolder* getAudioFileHolder();
```
Comment (bobtwinkles) [inline src/deluge/storage/multi_range/multi_wave_table_range.h:26]: Should have an `override` here

File: src/deluge/storage/multi_range/multisample_range.h [hunk 3]
```diff
@@ -29,6 +29,7 @@ class Cluster;
...
 
 	SampleHolderForVoice
 	    sampleHolder; // Has to be first variable, cos I do a sneaky optimization between both of MultiRange's children.
```
Comment (bobtwinkles) [inline src/deluge/storage/multi_range/multisample_range.h:35]: Can remove this comment

File: src/deluge/storage/multi_range/multi_wave_table_range.h [hunk 4]
```diff
@@ -23,6 +23,7 @@
...
 
 	WaveTableHolder
 	    waveTableHolder; // Has to be first variable, cos I do a sneaky optimization between both of MultiRange's children.
```
Comment (bobtwinkles) [inline src/deluge/storage/multi_range/multi_wave_table_range.h:29]: Can remove this comment

**Review by bobtwinkles — COMMENTED** [review]
Some stupid style nitpicks, but otherwise LGTM. Straightforward bug fix.

**Review by stellar-aria — APPROVED** [review]
@bobtwinkles issues are addressed and I don't see any other outstanding problems.

**Pull request merged**